### ISSUE TITLE
PR #1789 shards: Long tail - single-test failures (2 tests)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1938]
Part of #1931. **2 failing tests across 2 model(s)**, measured locally after the three landed fixes.

One test each. `MoreData_ShouldNotDegrade` on both, which also failed before any of the recent fixes.

### Failing tests

**VideoCLIP** (1)

- [x] `MoreData_ShouldNotDegrade`

**UFM** (1)

- [x] `MoreData_ShouldNotDegrade`

### Already fixed - do not redo

| Fix | Commit |
|---|---|
| Test probe derives from the model's declared contract (`GetArchitecture().GetInputShape()`) instead of a hardcoded `[1,4]` | `f73461b87` |
| `GlobalPoolingLayer` reduced channels+height instead of the spatial axes on NCHW | `f73461b87` |
| SSM layers published max sequence length as a fixed contract | `1239ef4a5` |

### Running these locally

```
dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj -c Debug -m:1 -f net10.0   --filter "FullyQualifiedName~<Model>Tests"
```

`-m:1` matters, and do not run two `dotnet test` invocations against this project at once - MSBuild fails copying `AiDotNetTests.dll` and the run reports nothing rather than failing loudly.

Parent: #1931

**Comment by ooples** [comment]
### UFM — fixed (`f045a926a`)

```
UFMTests: 27 tests, 27 passed
   MoreData_ShouldNotDegrade -> Passed
```

UFM had **no optimizer parameter at all**, so it took the framework default 1e-3 Adam. On identical data, one extra gradient step moved the loss from 23.86 to 100.66 while its 300k parameters shifted 0.0007 in L2 — divergence, not noise.

Its RAFT sibling in the same folder already documents this exact failure and remedy: dense-flow gradients overshoot at 1e-3, 1e-4 is the standard small optical-flow fine-tune rate, and `OpticalFlowBase`'s `maxGradNorm = 1.0` is close to a no-op under Adam so the LR is the effective lever. RAFT and FlowFormer both already expose an `optimizer` parameter — UFM was the outlier on both counts. It now matches its family rather than being tuned to a number that passes.

### VideoCLIP — still open

`2 iterations loss (0.7258) > 1 iterations loss (0.6702) + tolerance (0.0001)`

**Not the same cause.** VideoCLIP already runs AdamW at 1e-4 with gradient clipping at 1.0, consistent with its family (Donut 1e-4, SmolVLM 5e-5, SigLIP2 1e-5). Lowering it further would be tuning to pass rather than fixing anything.

Two things a fixer should weigh, neither of which I wanted to decide speculatively:

1. **Paper fidelity.** VideoCLIP (Xu et al. 2021) is *contrastive* pre-training over video-text pairs, but the model defaults to `MeanSquaredErrorLoss`. `ContrastiveLoss` and `NoiseContrastiveEstimationLoss` both exist in the repo. **However**, the sibling CLIP encoders default to MSE too, so this is a family-wide fidelity question rather than a VideoCLIP bug — changing one model in isolation would make it inconsistent with the rest.
2. **Tolerance scale.** `0.0001` against a loss of `0.67` demands the loss not rise by more than 0.015% across one extra step. The generator's own note (line 482) argues capping iterations beats relaxing `MoreDataTolerance` because a wide tolerance hides real divergence — worth re-reading before touching it.

### Correction worth recording

I initially diagnosed this whole issue as "the two networks train on different random data, so the comparison is noise". **That was wrong.** `ModelTestHelpers.CreateSeededRandom(int seed = 42)` means `CreateSeededRandom()` and `CreateSeededRandom(42)` return the same sequence — `input2` was always identical to `input`. The invariant is sound and was correctly reporting real divergence. Anyone reading the `rng1`/`rng2` pair in `MoreData_ShouldNotDegrade` will likely make the same misreading; the two RNGs look independent and are not.

**Comment by ooples** [comment]
## VideoCLIP paper/official-code comparison before the remaining fix

Current exact reproduction on PR head `345046366` / Tensors `0.120.6`:

`VideoCLIP.MoreData_ShouldNotDegrade` fails with one-step loss `0.6972740921289622` and two-step loss `0.725783555599588`; both parameter sets are finite and have effectively unchanged L2 norms (`20.7746579` vs `20.7746530`).

Original source comparison:

- Xu et al. §3.1 Eq. 3 applies **average pooling over the video and text token sequences** to obtain `z_v` and `z_t`; it does not select an EOS token.
- §3.2 Eqs. 4-5 use symmetric video-to-text/text-to-video InfoNCE over the **raw dot products** `z_v · z_t`, and §5.3 sets `τ = 1.0`.
- The official MMPT `MMFusionSeparate` returns raw average-pooled video/text states, and `MMContraLoss` forms both similarity matrices with matrix multiplication. The official config uses Adam `(0.9, 0.98)`, LR `5e-5`, 1000 warm-up updates, polynomial decay, clip norm `2.0`, and zero weight decay.

Sources: [ACL paper](https://aclanthology.org/2021.emnlp-main.544.pdf), [official model](https://github.com/facebookresearch/fairseq/blob/main/examples/MMPT/mmpt/models/mmfusion.py), [official contrastive loss](https://github.com/facebookresearch/fairseq/blob/main/examples/MMPT/mmpt/losses/nce.py), [official training config](https://github.com/facebookresearch/fairseq/blob/main/examples/MMPT/projects/retri/videoclip/how2.yaml).

Local `AiDotNet.Video.Understanding.VideoCLIP` now has the paper optimizer/schedule, but its representation geometry still differs:

- `EncodeVideo` and `EncodeText` L2-normalize their projected embeddings;
- text uses one EOS-position feature rather than average pooling the valid token sequence;
- `ComputeSimilarity` returns cosine similarity;
- the public temperature default is `0.07` rather than VideoCLIP's `1.0`.

I will correct those coupled VideoCLIP encoding/similarity defaults only. I will not change the generic test tolerance/iteration cap, redo UFM, or make another optimizer substitution.

**Comment by ooples** [comment]
## VideoCLIP — fixed and verified (`d9d65dfba`)

The remaining failure had two real model defects rather than an iteration/tolerance problem:

- VideoCLIP's explicit forward used private per-stage layer views. Deserialization replaced the framework-visible `Layers`, but those views stayed attached to constructor-era layers, so cloned training updated none of the published parameters. The views are now rebound after restore, and the learned token/positional embeddings are registered, persisted, and copied.
- The representation geometry disagreed with Xu et al.: text selected an EOS position, both towers L2-normalized, public similarity was cosine, and the default temperature was `0.07`. It now follows paper Eq. 1/3/4-5: stop-gradient at the pretrained spatial-backbone boundary, average-pool valid text states, use raw projected states and dot-product similarity, and default to `tau = 1.0`.

Primary references: [paper](https://aclanthology.org/2021.emnlp-main.544.pdf), [official model](https://github.com/facebookresearch/fairseq/blob/main/examples/MMPT/mmpt/models/mmfusion.py), [official loss](https://github.com/facebookresearch/fairseq/blob/main/examples/MMPT/mmpt/losses/nce.py), [official config](https://github.com/facebookresearch/fairseq/blob/main/examples/MMPT/projects/retri/videoclip/how2.yaml).

Final-tree focused verification, Release / net10.0 / `-m:1`:

```text
VideoCLIPTests.MoreData_ShouldNotDegrade
VideoExtendedIntegrationTests.VideoCLIP_CloneTraining_UpdatesPublishedParametersOnlyOnClone
Passed: 2, Failed: 0, Skipped: 0, Total: 2
```

Required full solution verification:

```text
dotnet build AiDotNet.sln -c Release --no-restore -m:1
Build succeeded. 3346 warnings, 0 errors.
Time Elapsed 00:03:13.75
```

This includes all target frameworks, including net471. UFM was already green in `f045a926a` and was not rerun.

**Comment by ooples** [comment]
### CI acceptance evidence

PR #1789 merge-tree run `30663655704` completed the VideoCLIP shard green:

- [ModelFamily — Generated Layers Vi–VM](https://github.com/ooples/AiDotNet/actions/runs/30663655704/job/91269215558) — **Success**

Together with the pushed 2/2 focused run and 0-error multi-target build in the prior evidence comment, this proves the remaining VideoCLIP item on the actual GitHub merge tree. Both #1938 model checkboxes are green.
